os/arch/arm/src/amebasmart/amebasmart_watchdog_lowerhalf.c

```diff
@@ -0,0 +1,485 @@
+	}
+	watchdog_init(priv->timeout_left);	// Re-init watchdog with remaining timeout
+	priv->int_mode = false;
+
+	/* Change to INT_MODE if handler exists */
+	if (handler) {
+		watchdog_irq_init((void *)priv->handler, (uint32_t)priv);	// INT_MODE enabled
+		priv->int_mode = true;
+	}
+
+	/* Re-start watchdog only if the status is WDT_START */
+	if (priv->wdt_status == WDT_START) {
+		watchdog_start();
+	}
+
+	return oldhandler;
+}
+
+/****************************************************************************
+ * Name: amebasmart_wdg_ioctl
+ *
+ * Description:
+ *   Any ioctl commands that are not recognized by the "upper-half" driver
+ *   are forwarded to the lower half driver through this method.
+ *
+ * Input Parameters:
+ *   lower - A pointer the publicly visible representation of the "lower-half"
+ *           driver state structure.
+ *   cmd   - The ioctl command value
+ *   arg   - The optional argument that accompanies the 'cmd'.  The
+ *           interpretation of this argument depends on the particular
+ *           command.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebasmart_wdg_ioctl(FAR struct watchdog_lowerhalf_s *lower, int cmd, unsigned long arg)
+{
+	struct amebasmart_wdg_lowerhalf_s *priv = (struct amebasmart_wdg_lowerhalf_s *)lower;
+	uint32_t time_elapse;
+	DEBUGASSERT(priv);
+	if (!priv) {
+		return -ENODEV;
+	}
+
+	switch (cmd) {
+	case WDIOC_PAUSE: {
+		/* Pause watchdog and get remaining timeout */
+		time_elapse = ((SYSTIMER_TickGet() - priv->start_tick - priv->on_pause_duration) * 31) / 1000;	// Get time elapse in ms
+		amebasmart_wdg_stop((struct watchdog_lowerhalf_s *)priv);
+		priv->on_pause_duration = SYSTIMER_TickGet();	// Record PAUSE start time
+		priv->timeout_left = priv->timeout_ms - time_elapse;	// Record remaining timeout
+		priv->wdt_status = WDT_PAUSE;
+	}
+	break;
+
+	case WDIOC_RESUME: {
+		/* Pre-assumption: the WDIOC_PAUSE can only be resumed by WDIOC_RESUME */
+		if (priv->wdt_status == WDT_PAUSE) {
+			/* Resume watchdog and init watchdog with remaining timeout */
+			priv->on_pause_duration = SYSTIMER_TickGet() - priv->on_pause_duration;
+			watchdog_init(priv->timeout_left);
+			if (priv->int_mode) {
+				watchdog_irq_init((void *)priv->handler, (uint32_t)priv);
+			}
+			watchdog_start();
+			priv->wdt_status = WDT_START;
+		}
+	}
+	break;
+
+	default: {
+		llvdbg("invalid cmd for wdg ioctl.\n");
+		return -EINVAL;
+	}
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebasmart_wdg_initialize
+ *
+ * Description:
+ *   Initialize and register the amebasmart watchdog. Watchdog timer is registered
+ *   as 'devpath'.
+ *
+ ****************************************************************************/
+int amebasmart_wdg_initialize(const char *devpath, uint32_t timeout_ms)
```

Comment:
Who calls this and timer init?

Reply:
> Who calls this and timer init?

Dear Mr Chang,
wdg init and timer init will be called in rtl8730e_boot.c if CONFIG_WATCHDOG or CONFIG_TIMER is enable.